### PR TITLE
Vanilla price modification

### PR DESCRIPTION
**Pull request by esztermarton**
editing vanilla prices to take in cost_of_carries and riskless_rate while keeping compatibility with version that takes in discount_factors.

### CODE REVIEW

File: tf_quant_finance/black_scholes/vanilla_prices.py [hunk 1]
```diff
@@ -64,21 +68,40 @@ def option_price(*,
     forwards: A real `Tensor` of any shape that broadcasts to the shape of
       `volatilities`. The forwards to maturity. Either this argument or the
       `spots` must be supplied but both must not be supplied.
+    discount_rates: An optional real `Tensor` of same dtype as the
+      `volatilities`. If not `None`, discount factors are calculated as e^(-rT),
+      where r are the discount rates, or risk free rates.
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/vanilla_prices.py]: Please add that either `discount_rates` or `discount_factors` must be supplied but not both. Please add the same comment for `discount_factors`.

File: tf_quant_finance/black_scholes/vanilla_prices.py [hunk 2]
```diff
@@ -64,21 +68,40 @@ def option_price(*,
...
+      Default value: `None`, equivalent to r = 0 and discount factors = 1 when
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/vanilla_prices.py:75]: `None` which maps to `-log(discount_factors) / expiries`

File: tf_quant_finance/black_scholes/vanilla_prices.py [hunk 3]
```diff
@@ -87,9 +110,18 @@ def option_price(*,
   Raises:
     ValueError: If both `forwards` and `spots` are supplied or if neither is
       supplied.
+    ValueError: If both `discount_rates` and `discount_factors` is supplied.
+    ValueError: If both `continuous_dividends` and `cost_of_carries` is
+      supplied.
   """
   if (spots is None) == (forwards is None):
     raise ValueError('Either spots or forwards must be supplied but not both.')
+  if discount_rates is not None and discount_factors is not None:
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/vanilla_prices.py]: please put parentheses as above to make it easier to read

File: tf_quant_finance/black_scholes/vanilla_prices.py [hunk 4]
```diff
@@ -87,9 +110,18 @@ def option_price(*,
...
+    raise ValueError('At most one of discount_rates and discount_factors may '
+                     'be supplied')
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/vanilla_prices.py:122]: Either discount_rates and discount_factors should be supplied but not both

File: tf_quant_finance/black_scholes/vanilla_prices_test.py [hunk 5]
```diff
@@ -134,6 +134,31 @@ def test_price_vol_and_expiry_scaling(self):
             forwards=forwards))
     self.assertArrayNear(base_prices, scaled_prices, 1e-10)
```
Comment (cyrilchim) [inline tf_quant_finance/black_scholes/vanilla_prices_test.py:137]: Please check both float32 and float64 via parametrized. Here and elsewhere. See, e.g., 
https://github.com/google/tf-quant-finance/blob/514236ce735ee2d6bdd9fb9981ce79da4dcaa1cd/tf_quant_finance/black_scholes/crr_binomial_tree_test.py#L30

**Review by cyrilchim — COMMENTED** [review]
Looks good, Rebeka! Can you please push the change? I will need to do the submission result via internal process.

File: tf_quant_finance/black_scholes/vanilla_prices.py (hunk 1 above)
Comment (esztermarton) [reply tf_quant_finance/black_scholes/vanilla_prices.py]: I've added that at most one of them can be supplied - I've maintained the default behaviour that it still works if neither is supplied!

File: tf_quant_finance/black_scholes/vanilla_prices.py (hunk 4 above)
Comment (esztermarton) [reply tf_quant_finance/black_scholes/vanilla_prices.py:122]: I've maintained the default behaviour that if neither is supplied it still works (assumed r = 0, discount factor = 1) - would you like me to change this default behaviour?

**Pull request merged**